### PR TITLE
backfill multi contract opportunity ids

### PR DESCRIPTION
**Pull request by muhammad-ammar**
**Description:** Backfill opportunity ids for enterprises with multiple contracts.

**JIRA:** [ENT-2887](https://openedx.atlassian.net/browse/ENT-2887)

### CODE REVIEW

File: ecommerce/enterprise/management/commands/backfill_opportunity_ids.py [hunk 1]
```diff
@@ -72,32 +83,52 @@ def read_csv(self, csv_path):
...
+            reader = csv.DictReader(csv_file)
+            for row in reader:
+                if row['ORDER_LINE_OFFER_TYPE'] == 'Voucher':
+                    data['coupons'][row['ORDER_LINE_COUPON_ID']].append(row['OPP_ID'])
+                elif row['ORDER_LINE_OFFER_TYPE'] in ('Site', 'User'):
+                    data['offers'][row['ORDER_LINE_OFFER_ID']].append(row['OPP_ID'])
+                else:
+                    data['ec_uuids'][row['ENTERPRISE_CUSTOMER_UUID']].append(row['OPP_ID'])
+
+        # condition the data so that at the end we have only one opportunity id for each coupon/offer
```
Comment (irfanuddinahmad) [inline ecommerce/enterprise/management/commands/backfill_opportunity_ids.py:102]: Given the work that we are doing for ENT-2999 ... do we still need to pick a "majority" opportunity?

Comment (muhammad-ammar) [reply ecommerce/enterprise/management/commands/backfill_opportunity_ids.py:102]: yes, that logic will remain same. Purpose of ENT-2999 is to override opportunity ids if ECS thinks they are incorrect.

File: ecommerce/enterprise/management/commands/backfill_opportunity_ids.py [hunk 2]
```diff
@@ -72,32 +83,52 @@ def read_csv(self, csv_path):
...
+                elif row['ORDER_LINE_OFFER_TYPE'] in ('Site', 'User'):
+                    data['offers'][row['ORDER_LINE_OFFER_ID']].append(row['OPP_ID'])
+                else:
+                    data['ec_uuids'][UUID(row['ENTERPRISE_CUSTOMER_UUID'])].append(row['OPP_ID'])
+
+        # condition the data so that at the end we have only one opportunity id for each coupon/offer
+        for __, category_data in data.items():
+            for category_object_id, opportunity_ids in category_data.items():
+                if len(opportunity_ids) > 1:
+                    most_common_opportunity_id, __ = Counter(opportunity_ids).most_common(1)[0]
```
Comment (irfanuddinahmad) [inline ecommerce/enterprise/management/commands/backfill_opportunity_ids.py:106]: Good!

File: ecommerce/enterprise/management/commands/backfill_opportunity_ids.py (hunk 1 above)
Comment (irfanuddinahmad) [reply ecommerce/enterprise/management/commands/backfill_opportunity_ids.py:102]: Would the override work as follows:
Check opportunity id with enrollment ... if not found, check in the associated order for the coupon or enterprise offer and find the opportunity id ... ?

Comment (irfanuddinahmad) [reply ecommerce/enterprise/management/commands/backfill_opportunity_ids.py:102]: ... if coupon is checked first ... I recommend that we donot add an opportunity id at the coupon level if there are multiple options ... that way ... if coupon is checked, an empty opportunity id is a signal that we need to check at the enrollment level

Comment (muhammad-ammar) [reply ecommerce/enterprise/management/commands/backfill_opportunity_ids.py:102]: The override work is completely separate from this backfilling work. For now we just need to set the opportunity ids for coupons and offers. I am not considering/thinking anything about overriding…Currently I am only focusing on backfilling which is just set the opportunity ids for coupons and offers.
Also override work is related to LMS not ecommerce

**Review by irfanuddinahmad — APPROVED** [review]
LGTM!

**Pull request merged**